File: solutions/clojure/cars-assemble/1/src/cars_assemble.clj
(ns cars-assemble)

(defn production-rate
  "Returns the assembly line's production rate per hour,
   taking into account its success rate"
  [speed]
  (cond
    (= speed 0)  (* 0.00 (* 221 speed))  ;; I understand you could just do 0.00, but I like the symmetry when reading
    (< speed 5)  (* 1.00 (* 221 speed))  ;; I understand you could just do (* 221 speed), but ^
    (< speed 9)  (* 0.90 (* 221 speed))
    (= speed 9)  (* 0.80 (* 221 speed))
    (= speed 10) (* 0.77 (* 221 speed))
  )
)

(defn working-items
  "Calculates how many working cars are produced per minute"
  [speed]
  (int (/ (production-rate speed) 60))
)

File: solutions/go/cars-assemble/1/cars_assemble.go
package cars

// CalculateWorkingCarsPerHour calculates how many working cars are
// produced by the assembly line every hour.
func CalculateWorkingCarsPerHour(productionRate int, successRate float64) float64 {
	return float64(productionRate) / 100.0 * successRate
}

// CalculateWorkingCarsPerMinute calculates how many working cars are
// produced by the assembly line every minute.
func CalculateWorkingCarsPerMinute(productionRate int, successRate float64) int {
	return int(CalculateWorkingCarsPerHour(productionRate, successRate) / 60.0)
}

// CalculateCost works out the cost of producing the given number of cars.
func CalculateCost(carsCount int) uint {
	return uint(95000 * (carsCount / 10) + 10000 * (carsCount % 10))
}
